## Choose and Justify Model Performance Indicator and Choice of Algorithm

In [1]:
# need a binary classifier (R or D for each county)
# logistic regression, support vector machine are options
# for deep learning, simple neural network

In [ ]:
# model performance indicator will be true positive rate and true negative rate (hit rates) vs. 
    # false negative and false positive (miss rates) because of 

## Generate Train and Test Sets

In [2]:
import pandas as pd
import numpy as np

In [8]:
# import dataset

data = pd.read_csv('features_v2.csv')
data

totalpop   income  service  office  drive  carpool  transit  \
0     10038388.0  56196.0    0.191   0.246  0.730    0.099    0.068   
1      5236393.0  55251.0    0.182   0.244  0.621    0.087    0.184   
2      4356362.0  54457.0    0.174   0.235  0.792    0.111    0.029   
3      4018143.0  54229.0    0.184   0.272  0.765    0.110    0.024   
4      2639042.0  43129.0    0.210   0.282  0.769    0.092    0.055   
...          ...      ...      ...     ...    ...      ...      ...   
3040       433.0  54306.0    0.104   0.279  0.475    0.097    0.000   
3041       901.0  33672.0    0.206   0.120  0.677    0.156    0.000   
3042       448.0  39375.0    0.109   0.041  0.522    0.156    0.000   
3043       565.0  36438.0    0.341   0.205  0.946    0.000    0.000   
3044       267.0  63250.0    0.102   0.139  0.886    0.000    0.000   

      workathome  meancommute  unemployment  perc_men  perc_white  \
0          0.051        0.300         0.100  0.492644    0.275897   
1          0.042        0.323         0.107  0.484541    0.438454   
2          0.033        0.282         0.075  0.497371    0.321501   
3          0.059        0.255         0.077  0.494297    0.585890   
4          0.043        0.299         0.100  0.485108    0.152371   
...          ...          ...           ...       ...         ...   
3040       0.226        0.320         0.009  0.515012    0.990854   
3041       0.038        0.170         0.039  0.488346    0.584337   
3042       0.199        0.195         0.040  0.497768    1.000000   
3043       0.000        0.166         0.000  0.522124    0.336000   
3044       0.036        0.116         0.051  0.509363    0.753000   

      perc_private_work  perc_citizen  party  
0              0.789211      0.602363      0  
1              0.839000      0.668383      0  
2              0.834835      0.572664      0  
3              0.824176      0.661961      1  
4              0.819000      0.576567      0  
...                 ...           ...    ...  
3040           0.693307      0.787529      1  
3041           0.744000      0.479467      1  
3042           0.544000      0.694196      1  
3043           0.519000      0.559292      0  
3044           0.729000      0.745318      1  

[3045 rows x 15 columns]

In [10]:
from sklearn.model_selection import train_test_split

X = data[data.columns[:-1]]
y = data['party']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Logistic Regression

### Implement Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

model.fit(X_train, y_train)

/Users/soujanyaibm.com/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

### Predict

In [13]:
y_pred = model.predict(X_test)

### Evaluation

In [14]:
model.score(X_test, y_test)

0.8702791461412152

In [19]:
    from sklearn.metrics import confusion_matrix
    confusion_matrix = confusion_matrix(y_test, y_pred)
    print(confusion_matrix)

[[ 24  73]
 [  6 506]]


# Neural Network

In [24]:
X.shape # 14 node input layer

(3045, 14)

In [ ]:
# 24 node dense layer with sigmoid activation
# sigmoid output layer

### Implement Neural Network

In [45]:
from keras.layers import Dense
from keras.layers import Input
from keras.models import Sequential
import keras.optimizers
import keras.metrics as metrics

model = Sequential()
model.add(Dense(14, input_dim=14, activation='relu'))
model.add(Dense(20, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[metrics.BinaryAccuracy(),metrics.AUC()])

model.fit(X_train, y_train)

### Predict & Evaluate

In [47]:
test_loss, test_acc = model.evaluate(X_test, y_test)[:2]

609/609 [==============================] - 0s 96us/step


In [48]:
print(test_loss, test_acc)

50.79996990335399 0.8210180401802063


# Conclusion

From the first preliminary run, logistic regression has 0.87 accuracy, and the neural network has 0.82 accuracy.

Next, I will try a logistic regression model and a neural network with all columns scaled between 0 and 1 and with state dummy variables.